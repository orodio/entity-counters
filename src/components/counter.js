import { html } from "@fp/html"
import { silence } from "@fp/silence"
import withCounter from "../sidecars/with-counter"
import counter from "../domains/counter"

export const Render = ({ counterId, count, inc = counter.inc, dec = counter.dec }) =>
  counterId == null
    ? html`
        <div>Loading...<//>
      `
    : html`
        <div>
          <div>
            <strong>${counterId}: <//>
            <span>${count}<//>
          <//>
          <div>
            <button onClick=${silence(_ => dec(counterId))}>dec<//>
            <button onClick=${silence(_ => inc(counterId))}>inc<//>
          <//>
        <//>
      `

export default withCounter(Render)
