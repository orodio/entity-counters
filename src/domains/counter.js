import { Entity } from "@fp/entity"
import { pick } from "@fp/pick"
import { getCount } from "./batch-fetch-counter"

const saveCountToApi = ({ id, count }) => Promise.resolve({ id, count })

export default Entity("counter", {
  withName: (id, label) => `${label}|${id}`,

  dependencies() {
    return {
      getCount,
      saveCount: saveCountToApi,
    }
  },

  broadcast: {
    count: state => pick(state, ["id", "count"]),
  },

  async init({ $$, Ok }, id) {
    const subs = new Set()
    const count = await $$.getCount(id)
    return Ok({ id, subs, count })
  },

  handleTell: {
    subscribe({ Ok, broadcast }, state, sub) {
      state.subs.add(sub)
      broadcast.count([sub], state)
      return Ok(state)
    },
    unsubscribe({ Ok }, state, sub) {
      state.subs.delete(sub)
      return Ok(state)
    },
    inc({ Ok, Continue, broadcast }, state, delta = 1) {
      state.count = state.count + delta
      broadcast.count(state.subs, state)
      return Ok(state, Continue("persist"))
    },
    dec({ Ok, Continue, broadcast }, state, delta = 1) {
      state.count = state.count - delta
      broadcast.count(state.subs, state)
      return Ok(state, Continue("persist"))
    },
  },

  handleContinue: {
    persist({ $$, Ok }, state) {
      $$.saveCount(pick(state, ["id", "count"]))
      return Ok(state)
    },
  },

  terminate(_ctx, _state, reason) {
    return reason
  },
})
