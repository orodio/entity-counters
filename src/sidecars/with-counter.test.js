import withCounter from "./with-counter"
import counter from "../domains/counter"
import { Entity } from "@fp/entity"
import { html } from "@fp/html"
import renderer from "react-test-renderer"

const sleep = ms => new Promise(resolve => setTimeout(() => resolve(ms), ms))

const Render = ({ counterId, count }) =>
  count == null
    ? html`
        <div>Loading...</div>
      `
    : html`
        <div>${counterId}: ${count}</div>
      `

test("withCounter", async () => {
  const ID = "foo"
  counter.init(ID, {
    debug: false,
    inject: {
      $$: {
        getCount: jest.fn(() => Promise.resolve(10)),
        saveCount: jest.fn(v => v),
      },
    },
  })

  const Counter = withCounter(Render)

  let c1 = renderer.create(
    html`
      <${Counter} counterId=${ID} />
    `
  )

  expect(c1.toJSON()).toMatchSnapshot("c1 loading")
  await sleep(50)
  expect(c1.toJSON()).toMatchSnapshot("c1 after subscribe")

  counter.inc(ID, 5)
  await sleep(50)
  expect(c1.toJSON()).toMatchSnapshot("c1 after inc")

  const c2 = renderer.create(
    html`
      <${Counter} counterId=${ID} />
    `
  )
  await sleep(50)
  expect(c2.toJSON()).toMatchSnapshot("c2 after subscribe")

  counter.dec(ID)
  await sleep(50)
  expect(c1.toJSON()).toMatchSnapshot("*c1* and c2 after dec")
  expect(c2.toJSON()).toMatchSnapshot("c1 and *c2* after dec")

  const t1 = await counter.dump(ID)
  expect(t1.subs.size).toBe(2)

  c1.unmount()
  c2.unmount()
  await sleep(50)
  const t2 = await counter.dump(ID)
  expect(t2.subs.size).toBe(0)
})
