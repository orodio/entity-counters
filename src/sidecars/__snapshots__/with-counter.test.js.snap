// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withCounter: *c1* and c2 after dec 1`] = `
<div>
  foo
  : 
  14
</div>
`;

exports[`withCounter: c1 after inc 1`] = `
<div>
  foo
  : 
  15
</div>
`;

exports[`withCounter: c1 after subscribe 1`] = `
<div>
  foo
  : 
  10
</div>
`;

exports[`withCounter: c1 and *c2* after dec 1`] = `
<div>
  foo
  : 
  14
</div>
`;

exports[`withCounter: c1 loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`withCounter: c2 after subscribe 1`] = `
<div>
  foo
  : 
  15
</div>
`;
